File: DifferentWaystoAddParentheses3.py
class Solution:
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        nums = []
        ops = []
        where = 0
        for i in range(len(input)):
            if input[i] in '+-*':
                nums.append(int(input[where:i]))
                ops.append(input[i])
                where = i + 1
        nums.append(int(input[where:]))

        dp = [[[] for _ in range(len(nums))] for i in range(len(nums))]
        for i in range(len(nums)):
            dp[i][i].append(nums[i])

        for i in range(len(nums) - 1, -1, -1):
            for j in range(i + 1, len(nums)):
                for k in range(i, j):
                    num1 = dp[i][k]
                    num2 = dp[k + 1][j]
                    for m in num1:
                        for n in num2:
                            if ops[k] == '+':
                                dp[i][j].append(m + n)
                            elif ops[k] == '-':
                                dp[i][j].append(m - n)
                            elif ops[k] == '*':
                                dp[i][j].append(m * n)
        return dp[0][len(nums) - 1]


if __name__ == '__main__':
    input = input()
    solution = Solution()
    print(solution.diffWaysToCompute(input))

File: DifferentWaystoAddParentheses3.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    vector<int> diffWaysToCompute(string input) {
    	
    	vector<int> result;
        
        int num = 0;
        for(int i = 0;i < input.length();i++) {
        	if(input[i] == '+' || input[i] == '-' || input[i] == '*') {
        		string p1 = input.substr(0, i);
        		string p2 = input.substr(i + 1, input.size() - i - 1);
        		vector<int> num1 = diffWaysToCompute(p1);
        		vector<int> num2 = diffWaysToCompute(p2);
        		
        		for(int j = 0;j < num1.size();j++) {
        			for(int k = 0;k < num2.size();k++) {
        				if(input[i] == '+') result.push_back(num1[j] + num2[k]);
        				else if(input[i] == '-') result.push_back(num1[j] - num2[k]);
        				else if(input[i] == '*') result.push_back(num1[j] * num2[k]);
					}
				}
			}else {
				num = num * 10 + (input[i] - '0');
			}
		}
		
		if(result.empty()) result.push_back(num);
		
		return result;
		
    }
};

int main() {
	
	string input;
	cin>>input;
	
	Solution *solution = new Solution();
	vector<int> result = solution->diffWaysToCompute(input);
	
	for(int i=0;i<result.size();i++) {
		cout<<result[i]<<" ";
	}
	
	return 0;
	
}
